src/NLog/Internal/CollectionExtensions.cs

```diff
@@ -68,10 +68,30 @@ internal static class CollectionExtensions
                         for (var j = 0; j < i; ++j)
                             filterLogEvents.Add(items[j]);
                     }
+
                     hasIgnoredLogEvents = true;
                 }
             }
+
             return filterLogEvents ?? (hasIgnoredLogEvents ? ArrayHelper.Empty<TItem>() : items);
         }
+
+        /// <summary>
+        /// Thread-safe getting value. Locks <paramref name="lockObject"/> only when needed.
+        /// </summary>
+        internal static bool TryGetValueThreadSafe<TKey, TValue>(this IDictionary<TKey, TValue> dict, TKey key, out TValue value, object lockObject)
```

Comment:
Please remove this method, and just do lookup and creation of factory method under the same lock

Reply:
So we won't have lock free resolves? Please note that DI will quickly involve in a lot of resolve calls.

Reply:
Only when loading config, and it will be performed with a single-thread.

Reply:
Some resolves are lazy? So that's not guaranteed from a single thread. This method is also used in the Resolve method for (user) targets etc.

Reply:
If the lookup happens for every LogEvent, then the Layout/Target is doing something wrong. Lock-free optimization should be implemented when it happens for every LogEvent.

Reply:
good point. Will also think about this, I will call it a night for today :)